pkg/ebpf/events_pipeline.go

```diff
@@ -242,14 +244,82 @@ func (t *Tracee) decodeEvents(outerCtx context.Context, sourceChan chan []byte)
 	return out, errc
 }
 
-// shouldProcessEvent decides whether or not to drop an event before further processing it
-func (t *Tracee) shouldProcessEvent(event trace.Event) bool {
+// computeScopes iterates through the scopes that do the filtering in user space, checking whether an event should be considered.
+// If it should not, it sets the respective offset to 0.
+// Finally it returns the bitmask of computed scopes.
+func (t *Tracee) computeScopes(event *trace.Event) uint64 {
 	eventID := events.ID(event.EventID)
+	origMatchedScopes := event.MatchedScopes
+	matchedScopes := event.MatchedScopes
+
+	for filterScope := range t.config.FilterScopes.Map() {
+		bitOffset := uint(filterScope.ID)
+
+		// Events submitted with matching scopes.
+		// The scope must have its bit cleared when it does not match.
+		if utils.HasBit(origMatchedScopes, bitOffset) {
+			// TODO: check if the event is traceable to some scope using a BPF map
+			if !filterScope.IsEventTraceable(eventID) {
```

Comment:
nit: there is an extra line break here.

This logic did not exist in the previous shouldProcessEvent() version.

What happens is the following: this function is called 2 times in the pipeline:

1. decodeEvents
1. deriveEvents

For derived events, let's pick "net_packet_ipv4" for example, the way it works is this:

1. it receives a "net_packet_ip_base" real event (coming from the kernel)
1. this event goes through decodeEvents
1. decodeEvents calls shouldProcessEvent() and it accepts the event.
1. this event goes through all the other pipeline stages
1. this event goes into deriveEvents and can or cannot be outputted (depends if the event is being traced) 
1. the derived event "net_packet_ipv4" will go through shouldProcessEvent() and, if not filtered, will be outputted

Now... in YOUR version.. what happens is the following:

1. it receives a "net_packet_ip_base" real event (coming from the kernel)
1. this event goes through decodeEvents
1. decodeEvents calls shouldProcessEvent() and it gets filtered: the reason why it gets filtered is that you added a logic "IsEventTraceable" that takes into consideration if the event was ever given to a scope. In the case of the derived event, the event is a dependency on the events to be filtered (since derivation did not occur yet). So, at this stage, you have the "net_packet_ip_base" event and you are filtering it because it wasn't selected in any scope.

You should go through all events given to any scope, loop through their event dependencies, and also accept those events, so they can be further derived. This will allow "net_packet_ip_base" to continue in the pipeline until it gets derived and filtered (since it wasn't picked in any scope).

There might be some gotchas in the middle (with the bit cleaning logic)

Reply:
Please include this patch: https://github.com/geyslan/tracee/pull/8 to this PR. It solves both issues: 

- derived events can now be filtered by scopes
- network events have matched_scopes correctly set

Example of usage:

```txt
$ sudo ./dist/tracee-ebpf --install-path /home/rafaeldtinoco/tracee/tmp --cache cache-type=mem --cache mem-cache-size=512 --output option:parse-arguments --capabilities bypass=false --output format:json --trace 1:event=net_packet_ipv4 --trace 1:net_packet_ipv4.args.dst="8.8.8.8" --trace 2:event=net_packet_ipv4 --trace 2:net_packet_ipv4.args.dst="1.1.1.1" --trace 3:event=openat --trace 3:openat.args.pathname="/etc/passwd" --trace 4:event=openat --trace 4:openat.args.pathname="/etc/shadow"
{"timestamp":1674015345045854703,"threadStartTime":1674015345043785470,"processorId":16,"processId":1576483,"cgroupId":6539,"threadId":1576483,"parentProcessId":1459289,"hostProcessId":1576483,"hostThreadId":1576483,"hostParentProcessId":1459289,"userId":1000,"mountNamespace":4026531841,"pidNamespace":4026531836,"processName":"ping","hostName":"fujitsu","containerId":"","containerImage":"","containerName":"","podName":"","podNamespace":"","podUID":"","podSandbox":false,"eventId":"2000","eventName":"net_packet_ipv4","matchedScopes":1,"argsNum":3,"returnValue":0,"stackAddresses":[0],"contextFlags":{"containerStarted":false},"args":[{"name":"src","type":"const char*","value":"192.168.100.2"},{"name":"dst","type":"const char*","value":"8.8.8.8"},{"name":"proto_ipv4","type":"trace.ProtoIPv4","value":{"version":4,"IHL":5,"TOS":0,"length":84,"id":51397,"flags":2,"fragOffset":0,"TTL":64,"protocol":"ICMPv4","checksum":15657,"srcIP":"192.168.100.2","dstIP":"8.8.8.8"}}]}
{"timestamp":1674015346047464836,"threadStartTime":1674015345043785470,"processorId":16,"processId":1576483,"cgroupId":6539,"threadId":1576483,"parentProcessId":1459289,"hostProcessId":1576483,"hostThreadId":1576483,"hostParentProcessId":1459289,"userId":1000,"mountNamespace":4026531841,"pidNamespace":4026531836,"processName":"ping","hostName":"fujitsu","containerId":"","containerImage":"","containerName":"","podName":"","podNamespace":"","podUID":"","podSandbox":false,"eventId":"2000","eventName":"net_packet_ipv4","matchedScopes":1,"argsNum":3,"returnValue":0,"stackAddresses":[0],"contextFlags":{"containerStarted":false},"args":[{"name":"src","type":"const char*","value":"192.168.100.2"},{"name":"dst","type":"const char*","value":"8.8.8.8"},{"name":"proto_ipv4","type":"trace.ProtoIPv4","value":{"version":4,"IHL":5,"TOS":0,"length":84,"id":51534,"flags":2,"fragOffset":0,"TTL":64,"protocol":"ICMPv4","checksum":15520,"srcIP":"192.168.100.2","dstIP":"8.8.8.8"}}]}
{"timestamp":1674015349234380912,"threadStartTime":1674015349232781700,"processorId":16,"processId":1576533,"cgroupId":6539,"threadId":1576533,"parentProcessId":1459289,"hostProcessId":1576533,"hostThreadId":1576533,"hostParentProcessId":1459289,"userId":1000,"mountNamespace":4026531841,"pidNamespace":4026531836,"processName":"cat","hostName":"fujitsu","containerId":"","containerImage":"","containerName":"","podName":"","podNamespace":"","podUID":"","podSandbox":false,"eventId":"257","eventName":"openat","matchedScopes":4,"argsNum":4,"returnValue":3,"stackAddresses":null,"contextFlags":{"containerStarted":false},"args":[{"name":"dirfd","type":"int","value":-100},{"name":"pathname","type":"const char*","value":"/etc/passwd"},{"name":"flags","type":"string","value":""},{"name":"mode","type":"mode_t","value":0}]}
```

Reply:
Awesome diving. Thanks @rafaeldtinoco.

Reply:
@rafaeldtinoco https://github.com/aquasecurity/tracee/actions/runs/3951089425